### PR TITLE
Memoize generation of rules for dune-package and meta files

### PR DESCRIPTION
**Pull request by snowleopard**
In our internal repo this reduces the Dune's memory consumption from 11GB to 5GB.

### CODE REVIEW

File: src/dune/install_rules.mli [hunk 1]
```diff
@@ -3,6 +3,16 @@ open Stdune
...
+    given super context and directory. Surprisingly, the current implementation
+    generates rules for {b all} directories of the project, which are filtered
+    out later in the [build_system.ml]. This function is called multiple times
+    for the same project (with different [dir]), so we memoize it to avoid
+    duplicating work. *)
+val meta_and_dune_package_rules : Super_context.t -> dir:Path.Build.t -> unit
+
+(* TODO: A seemingly more sensible approach for [meta_and_dune_package_rules]
+   is to only generate rules for the given directory, or stop taking [dir]
+   altogether and take [project] instead. *)
```
Comment (rgrinberg) [inline src/dune/install_rules.mli]: The latter approach seems to be worth trying. Wouldn't it be enough to just generate rules for the project's root dir to make it work?

Comment (aalekseyev) [reply src/dune/install_rules.mli]: Things can be cleaned up around here for sure. 
I think my preferred way would be to collect `meta_and_dune_package_rules_impl` and `install_rules` together into a single `Scheme.t` and leave the per-directory projection to `Scheme.evaluate`.

For this PR, cleanup is not the focus, so I think we should merge it.

Comment (aalekseyev) [reply src/dune/install_rules.mli]: Added to comment:

```
aalekseyev: actually I think we should just remove
   [meta_and_dune_package_rules] from the interface and have [gen_rules]
   do everything.
```

File: src/dune/install_rules.mli [hunk 2]
```diff
@@ -3,6 +3,16 @@ open Stdune
 val gen_rules :
   Super_context.t -> dir:Path.Build.t -> Build_system.Subdir_set.t
 
-val init_meta_and_dune_package : Super_context.t -> dir:Path.Build.t -> unit
+(** Generate rules for [.dune-package] and [META.<package-name>] files in a
+    given super context and directory. Surprisingly, the current implementation
+    generates rules for {b all} directories of the project, which are filtered
```
Comment (rgrinberg) [inline src/dune/install_rules.mli]: Where are they filtered out in build_system.ml? (Just for my own curiousity, I don't mean to block the PR).

Comment (aalekseyev) [reply src/dune/install_rules.mli]: It's done to ensure determinism. If you generate rules for dir `a`, but it happens to generate rules in `b`, then you have a choice: 1) use them; 2) ignore them; 3) report error.

(1) has a problem with determinism: rules for (b) end up depending on whether or not we forced (a) yet;
(3) was the plan originally, but it was hard to make all rules work so we did (2) as a "temporary" measure. Maybe we can clean up things so we switch to (3) later, but it's not a trivial fix.

Comment (aalekseyev) [reply src/dune/install_rules.mli]: Sorry, I read "why" where you said "where". Whoops. Let me try again...

Comment (aalekseyev) [reply src/dune/install_rules.mli]: This line:
```
Rules.find rules_produced dir
```

effectively filters the rules to just the ones produced for `dir`.
